File: Web/src/schema/bridge-rpc.ts
import type { BaseObject, ObjectId, ScopeConfig } from "./base";

/* ============================================================================
 * bridge-rpc.ts — the contract between the React UI and the native side.
 *
 * Today's bridge (Web/src/bridge.ts) carries only navigate/theme/insets. This
 * defines the two channels everything else needs:
 *
 *   1. RPC      JS calls a BridgeApi method, awaits a typed result.
 *   2. Events   Swift pushes a BridgeEvent into the page (alerts, streaming).
 *
 * Typed against the Object schema, so both sides share one source of truth.
 * The methods are grouped by surface; each maps to a local service from the UI
 * spec (vault FS, index/search, embeddings, LLM/agent, connectors, store).
 * ========================================================================== */

/* ---- transport envelopes ----------------------------------------------- */

export interface RpcCall<M extends keyof BridgeApi> {
  id: string; // correlation id; the result echoes it
  method: M;
  params: Parameters<BridgeApi[M]>[0];
}
export interface RpcResult<M extends keyof BridgeApi> {
  id: string;
  ok: boolean;
  result?: Awaited<ReturnType<BridgeApi[M]>>;
  error?: { code: string; message: string };
}

/* ---- shared shapes ----------------------------------------------------- */

export interface FileNode {
  path: string;
  name: string;
  kind: "folder" | "file";
  ext?: string;
  children?: FileNode[];
}

export interface ObjectQuery {
  type?: string; // a type or any of its subtypes (lineage-aware on the native side)
  clusterId?: ObjectId;
  text?: string;
  limit?: number;
}

export type SearchResultObject = BaseObject; // type === "search.result"
export type MoveObject = BaseObject; // type === "move"
export type ClusterObject = BaseObject; // type === "cluster" | "cluster.artifact"
export type LoopObject = BaseObject; // type === "loop"
export type AlertObject = BaseObject; // type === "alert"
export type SourceObject = BaseObject; // type === "source"

export type GateAction = "accept" | "adjust" | "approve" | "hold" | "simulate-reply";

/* ---- the API the native side implements -------------------------------- */

export interface BridgeApi {
  /* objects: the generic CRUD over any primitive */
  "objects.get": (p: { id: ObjectId }) => Promise<BaseObject | null>;
  "objects.put": (p: { object: BaseObject }) => Promise<BaseObject>; // validates, persists, (re)indexes
  "objects.delete": (p: { id: ObjectId }) => Promise<{ deleted: boolean }>;
  "objects.query": (p: ObjectQuery) => Promise<BaseObject[]>;

  /* vault: live Obsidian directory. Notes are pointer-storage; writes hit the .md */
  "vault.tree": (p: { root?: string }) => Promise<FileNode[]>;
  "vault.readNote": (p: { path: string }) => Promise<{ object: BaseObject; markdown: string }>;
  "vault.writeNote": (p: { path: string; markdown: string }) => Promise<BaseObject>; // writes .md, re-derives metadata
  "vault.move": (p: { from: string; to: string }) => Promise<{ ok: boolean }>;
  "vault.rename": (p: { path: string; name: string }) => Promise<{ ok: boolean }>;
  "vault.createFolder": (p: { path: string }) => Promise<{ ok: boolean }>;
  "vault.delete": (p: { path: string }) => Promise<{ ok: boolean }>;

  /* search + retrieval: one hybrid path, two callers (user UI and model) */
  "search.hybrid": (p: { query: string; scope?: ScopeConfig; k?: number }) => Promise<SearchResultObject[]>;
  "search.retrieve": (p: { query: string; scope?: ScopeConfig; k?: number }) => Promise<{
    chunks: Array<{ text: string; sourceId: ObjectId; score: number }>;
  }>;

  /* index + embeddings */
  "index.reindex": (p: { sourceId?: ObjectId; objectId?: ObjectId }) => Promise<{ jobId: string }>;
  "index.status": (p: { sourceId?: ObjectId }) => Promise<{ status: string; done: number; total: number }>;
  "embeddings.setProvider": (p: { provider: "local" | "openai" | "google"; model: string }) => Promise<{ ok: boolean }>;

  /* moves: the spine + correct-and-recompute + gated execution */
  "moves.propose": (p: { loopId?: ObjectId; seed?: ObjectId }) => Promise<MoveObject>;
  "moves.act": (p: { moveId: ObjectId; nodeId: string; action: GateAction; feedback?: string }) => Promise<MoveObject>;
  "moves.execute": (p: { moveId: ObjectId }) => Promise<{ move: MoveObject; exit: BaseObject }>;

  /* clusters: the home board + auto-clustering + artifacts */
  "clusters.list": () => Promise<ClusterObject[]>;
  "clusters.create": (p: { members: ObjectId[]; kind?: string; title?: string }) => Promise<ClusterObject>;
  "clusters.autocluster": () => Promise<ClusterObject[]>; // a structuring pass; returns proposals (confidence-gated)
  "clusters.setMembers": (p: { clusterId: ObjectId; add?: ObjectId[]; remove?: ObjectId[] }) => Promise<ClusterObject>;
  "clusters.artifacts": (p: { clusterId: ObjectId }) => Promise<ClusterObject[]>;

  /* loops: Carrying */
  "loops.list": (p: { clusterId?: ObjectId; segment?: "all" | "needs-you" | "waiting" | "fires" }) => Promise<LoopObject[]>;
  "loops.setState": (p: { loopId: ObjectId; state: string }) => Promise<LoopObject>;

  /* alerts: raised by the model tool OR auto-emitted on `fire` tagging */
  "alerts.raise": (p: { target: ObjectId; severity: "info" | "warn" | "fire"; headline: string; context?: string; trigger: "model" | "auto" }) => Promise<AlertObject>;
  "alerts.snooze": (p: { alertId: ObjectId; until: string }) => Promise<AlertObject>;
  "alerts.list": () => Promise<AlertObject[]>;

  /* ai sessions: scoped chat. Streaming arrives via events, not the return. */
  "ai.start": (p: { scope?: ScopeConfig }) => Promise<{ sessionId: ObjectId }>;
  "ai.send": (p: { sessionId: ObjectId; message: string; pinned?: ObjectId[] }) => Promise<{ accepted: boolean }>;

  /* connectors / sources */
  "sources.list": () => Promise<SourceObject[]>;
  "sources.reconnect": (p: { sourceId: ObjectId }) => Promise<{ ok: boolean }>;
  "sources.sweep": (p: { sourceId: ObjectId }) => Promise<{ jobId: string }>;
}

/* ---- the events the native side pushes (Swift -> JS) -------------------- */

export interface BridgeEvents {
  /** The Reactive banner. Same shape whether model-raised or auto-emitted. */
  alert: AlertObject;
  /** Streaming chat tokens for an ai.session. */
  "chat.token": { sessionId: ObjectId; token: string };
  "chat.done": { sessionId: ObjectId; session: BaseObject };
  /** Reindex / embedding progress for the Sources + Vault status UI. */
  "index.progress": { sourceId?: ObjectId; done: number; total: number; status: string };
  /** Obsidian (or anything) changed a vault file underneath us. */
  "vault.changed": { path: string; kind: "created" | "modified" | "deleted" };
  /** A loop changed state (Carrying live-updates). */
  "loop.updated": LoopObject;
  /** Connector sweep heartbeat (the sidebar "listening / swept" status). */
  "sweep.status": { sourceId: ObjectId; status: string; at: string };
}

export type BridgeEventName = keyof BridgeEvents;
export interface BridgeEvent<E extends BridgeEventName = BridgeEventName> {
  event: E;
  payload: BridgeEvents[E];
}

File: Documents/Handoff/bridge/BridgeRPC.swift
import Foundation

// =============================================================================
// BridgeRPC.swift — the native side of the bridge contract.
//
// The authoritative contract is bridge-rpc.ts (typechecked). This is the Swift
// mirror: the envelopes the WKWebView message handler decodes, the method and
// event names, and the service protocol the app implements. Bodies stay generic
// (JSONValue) at this layer; each handler decodes its own params/result struct.
//
// Wire it to the existing WebBridge: today JS posts only {type:"ready"}. Add a
// second message type {type:"rpc", call: RpcCall} and reply by evaluating
// `window.__steward.resolve(id, result)`; push events via
// `window.__steward.emit(event, payload)`.
// =============================================================================

public struct RpcCall: Codable, Sendable {
    public let id: String
    public let method: String
    public let params: JSONValue
}

public struct RpcResult: Codable, Sendable {
    public let id: String
    public let ok: Bool
    public var result: JSONValue?
    public var error: RpcError?
}

public struct RpcError: Codable, Sendable {
    public let code: String
    public let message: String
}

/// Names mirror bridge-rpc.ts `BridgeApi`. Kept as a flat namespace of strings
/// so the dispatcher can switch on them.
public enum BridgeMethod: String, Sendable {
    case objectsGet = "objects.get"
    case objectsPut = "objects.put"
    case objectsDelete = "objects.delete"
    case objectsQuery = "objects.query"
    case vaultTree = "vault.tree"
    case vaultReadNote = "vault.readNote"
    case vaultWriteNote = "vault.writeNote"
    case vaultMove = "vault.move"
    case vaultRename = "vault.rename"
    case vaultCreateFolder = "vault.createFolder"
    case vaultDelete = "vault.delete"
    case searchHybrid = "search.hybrid"
    case searchRetrieve = "search.retrieve"
    case indexReindex = "index.reindex"
    case indexStatus = "index.status"
    case embeddingsSetProvider = "embeddings.setProvider"
    case movesPropose = "moves.propose"
    case movesAct = "moves.act"
    case movesExecute = "moves.execute"
    case clustersList = "clusters.list"
    case clustersCreate = "clusters.create"
    case clustersAutocluster = "clusters.autocluster"
    case clustersSetMembers = "clusters.setMembers"
    case clustersArtifacts = "clusters.artifacts"
    case loopsList = "loops.list"
    case loopsSetState = "loops.setState"
    case alertsRaise = "alerts.raise"
    case alertsSnooze = "alerts.snooze"
    case alertsList = "alerts.list"
    case aiStart = "ai.start"
    case aiSend = "ai.send"
    case sourcesList = "sources.list"
    case sourcesReconnect = "sources.reconnect"
    case sourcesSweep = "sources.sweep"
}

/// Push events (Swift -> JS). Mirrors bridge-rpc.ts `BridgeEvents`.
public enum BridgeEvent: Sendable {
    case alert(StewardObject)
    case chatToken(sessionId: String, token: String)
    case chatDone(sessionId: String, session: StewardObject)
    case indexProgress(sourceId: String?, done: Int, total: Int, status: String)
    case vaultChanged(path: String, kind: String) // created | modified | deleted
    case loopUpdated(StewardObject)
    case sweepStatus(sourceId: String, status: String, at: String)

    public var name: String {
        switch self {
        case .alert: return "alert"
        case .chatToken: return "chat.token"
        case .chatDone: return "chat.done"
        case .indexProgress: return "index.progress"
        case .vaultChanged: return "vault.changed"
        case .loopUpdated: return "loop.updated"
        case .sweepStatus: return "sweep.status"
        }
    }
}

/// What the app implements. One actor backs the store; vault/index/embeddings/
/// LLM/connectors are services it calls. Representative methods only; the full
/// surface is in bridge-rpc.ts.
public protocol BridgeService: Sendable {
    func objectsGet(id: String) async throws -> StewardObject?
    func objectsPut(_ object: StewardObject) async throws -> StewardObject
    func objectsQuery(type: String?, clusterId: String?, text: String?, limit: Int?) async throws -> [StewardObject]

    func vaultTree(root: String?) async throws -> JSONValue
    func vaultReadNote(path: String) async throws -> (object: StewardObject, markdown: String)
    func vaultWriteNote(path: String, markdown: String) async throws -> StewardObject

    func searchHybrid(query: String, scope: JSONValue?, k: Int?) async throws -> [StewardObject]

    func movesAct(moveId: String, nodeId: String, action: String, feedback: String?) async throws -> StewardObject
    func movesExecute(moveId: String) async throws -> (move: StewardObject, exit: StewardObject)

    func alertsRaise(target: String, severity: String, headline: String, context: String?, trigger: String) async throws -> StewardObject

    /// Subscribe to push events; the implementation calls back on the main actor
    /// to evaluate `window.__steward.emit(...)` in the WKWebView.
    func subscribe(_ onEvent: @escaping @Sendable (BridgeEvent) -> Void)
}
